context/src/source.rs

```diff
@@ -55,27 +76,7 @@ impl SourceSegmentHolder for Source<'_> {
     }
 }
 
-impl<'s> SourceCode for &'s Source<'_> {
```

Comment:
Why? This seems unrelated to move it in the CLI from the original goal of the PR, simply remove the `&'s` part.

Reply:
The dependency of a diagnostic library such as miette should only be contained by top-level crates. I think that the use of miette in the `context` crate kinda forces the higher crates to depend on it. For example, will we really need to depend on a diagnostic library for a LSP implementation ? ik this is kinda out of scope but i took the advantage of the fact that i was refactoring the cli to fix this useless dependency

Reply:
Dependencies are not transitive in Rust, and miette was designed to have its core types in the default crate, and expose the fancy graphical stuff behind a feature flag. I agree that it did break the original "no external dependency" contract, but this concession was made deliberately, in my opinion, to avoid having to create too many `Source`s types that are strictly equivalents. On the other hand, something like `Diagnostic` was completely separated from `ParseError`, as their respective representations differ.

Reply:
> For example, will we really need to depend on a diagnostic library for a LSP implementation ?

An editor integration also works with the context, but does not indeed necessary use miette. If this is really a concern, put the implementation behind a feature flag.

> i was refactoring the cli to fix this useless dependency

The dependency is still declared in the `Cargo.toml` file :p